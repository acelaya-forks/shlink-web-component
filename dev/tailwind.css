@import 'tailwindcss' prefix(tw) important;
@source '../node_modules/@shlinkio/shlink-frontend-kit';
@import '@shlinkio/shlink-frontend-kit/tailwind.preset.css';
