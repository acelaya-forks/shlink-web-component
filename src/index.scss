@use '../node_modules/@shlinkio/shlink-frontend-kit/dist/base';
@use './tags/react-tag-autocomplete';
@use 'leaflet/dist/leaflet.css';

dialog .leaflet-top .leaflet-control {
  margin-top: 60px;
}
