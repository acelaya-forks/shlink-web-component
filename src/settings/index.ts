import { createContext, useContext } from 'react';
import type { QrCodeSettings, Settings, VisitsListSettings } from './types';

export const defaultQrCodeSettings: QrCodeSettings = {
  size: 300,
  margin: 0,
  color: '#000000',
  bgColor: '#ffffff',
  errorCorrection: 'L',
  format: 'png',
} as const;

Object.freeze(defaultQrCodeSettings);

export const defaultVisitsListColumns: Required<VisitsListSettings['columns']> = {
  potentialBot: true,
  date: true,
  country: true,
  region: false,
  city: true,
  browser: true,
  os: true,
  userAgent: false,
  referer: true,
  visitedUrl: true,
} as const;

Object.freeze(defaultVisitsListColumns);

const defaultSettings: Settings = {
  realTimeUpdates: {
    enabled: true,
  },
  shortUrlCreation: {
    validateUrls: false,
  },
  visits: {
    defaultInterval: 'last30Days',
  },
  visitsList: {
    columns: defaultVisitsListColumns,
  },
  shortUrlsList: {
    defaultOrdering: {
      field: 'dateCreated',
      dir: 'DESC',
    },
  },
  qrCodes: defaultQrCodeSettings,
};

const SettingsContext = createContext(defaultSettings);

export const { Provider: SettingsProvider } = SettingsContext;

export const useSettings = (): Settings => useContext(SettingsContext) ?? defaultSettings;

export function useSetting<Setting extends keyof Settings>(settingName: Setting): Settings[Setting];
export function useSetting<Setting extends keyof Settings, Default extends NonNullable<Settings[Setting]>>(
  settingName: Setting,
  fallbackValue: Default,
): NonNullable<Settings[Setting]>;
export function useSetting<Setting extends keyof Settings, Default extends Settings[Setting]>(
  settingName: Setting,
  fallbackValue?: Default,
) {
  const settings = useSettings();
  return settings[settingName] ?? fallbackValue;
}

export { ShlinkWebSettings } from './components/ShlinkWebSettings';
export * from './types';
