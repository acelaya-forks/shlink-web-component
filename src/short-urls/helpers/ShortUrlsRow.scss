@use '../../../node_modules/@shlinkio/shlink-frontend-kit/dist/base';
@use '../../utils/mixins/vertical-align';

@mixin text-ellipsis() {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.short-urls-row__cell.short-urls-row__cell {
  vertical-align: middle !important;
}

.short-urls-row__cell--break {
  word-break: break-all;
}

.short-urls-row__cell--indivisible {
  @media (min-width: base.$lgMin) {
    white-space: nowrap;
  }
}

.short-urls-row__short-url-wrapper {
  @media (max-width: base.$mdMax) {
    word-break: break-all;
  }

  @media (min-width: base.$lgMin) {
    @include text-ellipsis();

    vertical-align: bottom;
    display: inline-block;
    max-width: 18rem;
  }
}

.short-urls-row__copy-hint {
  @include vertical-align.vertical-align(translateX(10px));

  box-shadow: 0 3px 15px rgb(0 0 0 / .25);

  @media (max-width: base.$responsiveTableBreakpoint) {
    @include vertical-align.vertical-align(translateX(calc(-100% - 20px)));
  }
}
