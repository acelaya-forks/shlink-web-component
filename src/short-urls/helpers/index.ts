import type { ShlinkCreateShortUrlData, ShlinkShortUrl, ShlinkShortUrlIdentifier } from '../../api-contract';
import { DEFAULT_DOMAIN } from '../../domains/data';
import type { ShortUrlCreationSettings } from '../../settings';
import type { OptionalString } from '../../utils/helpers';

export const shortUrlMatches = (shortUrl: ShlinkShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (domain === undefined || domain === null) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShlinkShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

// FIXME This should return ShlinkEditShortUrlData
export const shortUrlDataFromShortUrl = (
  shortUrl?: ShlinkShortUrl,
  settings?: ShortUrlCreationSettings,
): ShlinkCreateShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls && {
      android: shortUrl.deviceLongUrls.android ?? undefined,
      ios: shortUrl.deviceLongUrls.ios ?? undefined,
      desktop: shortUrl.deviceLongUrls.desktop ?? undefined,
    },
    validateUrl,
  };
};

/**
 * Converts a short code into a valid URL param, replacing bars from multi-segment slugs with double underscore
 */
export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', '__');

/**
 * Converts a URL param representing a short code into its corresponding short code, by replacing double underscores
 * (if any) with bars, which means it's a multi-segment slug
 */
export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll('__', '/');

/**
 * String representation of a short URL, so that it can be used as query param
 */
export const shortUrlToQuery = ({ domain, shortCode }: ShlinkShortUrlIdentifier): string =>
  `${domain ?? DEFAULT_DOMAIN}__${urlEncodeShortCode(shortCode)}`;

/**
 * String representation of a short URL, so that it can be used as query param
 */
export const queryToShortUrl = (shortUrlQuery: string): ShlinkShortUrlIdentifier => {
  // Split in two segments only. The second one may also contain the split placeholder if it is a multi-segment slug
  const [domain, shortCode] = shortUrlQuery.split(/__(.+)/);
  if (!shortCode) {
    throw new Error(`It was not possible to parse domain and short code from "${shortUrlQuery}"`);
  }

  return {
    domain: domain === DEFAULT_DOMAIN ? null : domain,
    shortCode: urlDecodeShortCode(shortCode),
  };
};
