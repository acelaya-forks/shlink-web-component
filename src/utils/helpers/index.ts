import { range } from '@shlinkio/data-manipulation';
import type { SyntheticEvent } from 'react';
import { useCallback } from 'react';

export type OptionalString = string | null | undefined;

/**
 * Wraps an event handler so that it calls e.preventDefault() before invoking the event handler
 */
export const usePreventDefault = <Event extends SyntheticEvent = SyntheticEvent>(handler: (e: Event) => void) =>
  useCallback((e: Event) => {
    e.preventDefault();
    handler(e);
  }, [handler]);

export const rangeOf = <T>(size: number, mappingFn: (value: number) => T, startAt = 1): T[] =>
  range(startAt, size + 1).map(mappingFn);

export type Empty = null | undefined | '' | never[];

const isEmpty = (value: NonNullable<any>): boolean => (
  (Array.isArray(value) && value.length === 0)
  || (typeof value === 'string' && value === '')
  || (typeof value === 'object' && Object.keys(value).length === 0)
);

export const hasValue = <T>(value: T | Empty): value is T => value !== undefined && value !== null && !isEmpty(value);

export const nonEmptyStringOrNull = <T extends string>(value: T): T | null => (!value ? null : value);

export type BooleanString = 'true' | 'false';

export const parseBooleanToString = (value: boolean): BooleanString => (value ? 'true' : 'false');

export const parseOptionalBooleanToString = (value?: boolean): BooleanString | undefined => (
  value === undefined ? undefined : parseBooleanToString(value)
);
