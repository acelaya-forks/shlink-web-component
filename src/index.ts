import { bottle } from './container';
import { createShlinkWebComponent } from './ShlinkWebComponent';
import { fixLeafletIcons } from './utils/helpers/leaflet';

// This overwrites icons used for leaflet maps, fixing some issues caused by webpack while processing the CSS
fixLeafletIcons();

export const ShlinkWebComponent = createShlinkWebComponent(bottle);

export type ShlinkWebComponentType = typeof ShlinkWebComponent;

export type { ShlinkWebComponentProps } from './ShlinkWebComponent';

export type { TagColorsStorage } from './utils/services/TagColorsStorage';

export { ShlinkSidebarToggleButton } from './sidebar/ShlinkSidebarToggleButton';
export type { ShlinkSidebarToggleButtonProps } from './sidebar/ShlinkSidebarToggleButton';
export { ShlinkSidebarVisibilityProvider } from './sidebar/ShlinkSidebarVisibilityProvider';
