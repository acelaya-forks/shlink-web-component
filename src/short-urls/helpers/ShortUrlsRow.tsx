import { faArrowsSplitUpAndLeft as rulesIcon } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { TimeoutToggle } from '@shlinkio/shlink-frontend-kit';
import { CopyToClipboardButton, Table } from '@shlinkio/shlink-frontend-kit';
import { useEffect, useRef } from 'react';
import { ExternalLink } from 'react-external-link';
import type { ShlinkShortUrl } from '../../api-contract';
import type { FCWithDeps } from '../../container/utils';
import { componentFactory, useDependencies } from '../../container/utils';
import { useSetting } from '../../settings';
import { Time } from '../../utils/dates/Time';
import type { ColorGenerator } from '../../utils/services/ColorGenerator';
import { useShortUrlsQuery } from './hooks';
import { ShortUrlDetailLink } from './ShortUrlDetailLink';
import type { ShortUrlsRowMenuType } from './ShortUrlsRowMenu';
import { ShortUrlStatus } from './ShortUrlStatus';
import { ShortUrlVisitsCount } from './ShortUrlVisitsCount';
import { Tags } from './Tags';

type ShortUrlsRowProps = {
  onTagClick?: (tag: string) => void;
  shortUrl: ShlinkShortUrl;
};

type ShortUrlsRowDeps = {
  ShortUrlsRowMenu: ShortUrlsRowMenuType,
  ColorGenerator: ColorGenerator,
  useTimeoutToggle: TimeoutToggle,
};

export type ShortUrlsRowType = typeof ShortUrlsRow;

const ShortUrlsRow: FCWithDeps<ShortUrlsRowProps, ShortUrlsRowDeps> = ({ shortUrl, onTagClick }) => {
  const { ShortUrlsRowMenu, ColorGenerator: colorGenerator, useTimeoutToggle } = useDependencies(ShortUrlsRow);
  // eslint-disable-next-line react-compiler/react-compiler
  const [active, setActive] = useTimeoutToggle({ initialValue: false, delay: 500 });
  const isFirstRun = useRef(true);
  const [{ excludeBots }] = useShortUrlsQuery();
  const visits = useSetting('visits');
  const doExcludeBots = excludeBots ?? visits?.excludeBots;

  useEffect(() => {
    // TODO In order to highlight short URLs when new visits occur, there should be an action adding the short URLs to
    //  highlight to a list, and taking care of removing them from that list after the timeout.
    if (!isFirstRun.current) {
      setActive();
    }

    isFirstRun.current = false;
  }, [shortUrl.visitsSummary?.total, shortUrl.visitsSummary?.nonBots, shortUrl.visitsCount, setActive]);

  return (
    <Table.Row className="relative">
      <Table.Cell className="whitespace-nowrap" columnName="Created at:">
        <Time date={shortUrl.dateCreated} />
      </Table.Cell>
      <Table.Cell columnName="Short URL:">
        <span className="lg:whitespace-nowrap inline-flex items-center gap-x-2">
          <ExternalLink href={shortUrl.shortUrl} className="max-md:break-all lg:truncate max-w-72" />
          <CopyToClipboardButton text={shortUrl.shortUrl} />
        </span>
      </Table.Cell>
      <Table.Cell className="break-all" columnName={`${shortUrl.title ? 'Title' : 'Long URL'}:`}>
        <ExternalLink href={shortUrl.longUrl}>{shortUrl.title ?? shortUrl.longUrl}</ExternalLink>
      </Table.Cell>
      {shortUrl.title && (
        <Table.Cell className="break-all [&]:lg:hidden" columnName="Long URL:">
          <ExternalLink href={shortUrl.longUrl} />
        </Table.Cell>
      )}
      <Table.Cell columnName="Tags:">
        <Tags tags={shortUrl.tags} colorGenerator={colorGenerator} onTagClick={onTagClick} />
      </Table.Cell>
      <Table.Cell className="lg:text-right" columnName="Visits:">
        <ShortUrlVisitsCount
          visitsCount={(
            doExcludeBots ? shortUrl.visitsSummary?.nonBots : shortUrl.visitsSummary?.total
          ) ?? shortUrl.visitsCount ?? 0}
          shortUrl={shortUrl}
          active={active}
          asLink
        />
      </Table.Cell>
      <Table.Cell columnName="Status:" className="max-lg:border-none">
        <div className="inline-flex gap-2">
          <ShortUrlStatus shortUrl={shortUrl} />
          {shortUrl.hasRedirectRules && (
            <ShortUrlDetailLink
              asLink
              shortUrl={shortUrl}
              suffix="redirect-rules"
              title="This short URL has dynamic redirect rules"
            >
              <FontAwesomeIcon icon={rulesIcon} />
            </ShortUrlDetailLink>
          )}
        </div>
      </Table.Cell>
      <Table.Cell className="text-right max-lg:absolute max-lg:top-1 max-lg:right-1 max-lg:p-0">
        <ShortUrlsRowMenu shortUrl={shortUrl} />
      </Table.Cell>
    </Table.Row>
  );
};

export const ShortUrlsRowFactory = componentFactory(
  ShortUrlsRow,
  ['ShortUrlsRowMenu', 'ColorGenerator', 'useTimeoutToggle'],
);
