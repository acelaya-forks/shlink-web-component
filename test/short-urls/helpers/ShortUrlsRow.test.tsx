import { Table } from '@shlinkio/shlink-frontend-kit/tailwind';
import { screen } from '@testing-library/react';
import { fromPartial } from '@total-typescript/shoehorn';
import { addDays, formatISO, subDays } from 'date-fns';
import { MemoryRouter } from 'react-router';
import type { ShlinkShortUrl, ShlinkShortUrlMeta } from '../../../src/api-contract';
import type { Settings } from '../../../src/settings';
import { SettingsProvider } from '../../../src/settings';
import { ShortUrlsRowFactory } from '../../../src/short-urls/helpers/ShortUrlsRow';
import { now, parseDate } from '../../../src/utils/dates/helpers/date';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';
import { colorGeneratorMock } from '../../utils/services/__mocks__/ColorGenerator.mock';

interface SetUpOptions {
  title?: string | null;
  tags?: string[];
  meta?: ShlinkShortUrlMeta;
  settings?: Partial<Settings>;
  search?: string;
  hasRedirectRules?: boolean;
}

describe('<ShortUrlsRow />', () => {
  const timeoutToggle = vi.fn(() => true);
  const useTimeoutToggle = vi.fn(() => [false, timeoutToggle]);
  const shortUrl: ShlinkShortUrl = {
    shortCode: 'abc123',
    shortUrl: 'https://s.test/abc123',
    longUrl: 'https://foo.com/bar',
    dateCreated: formatISO(parseDate('2018-05-23 18:30:41', 'yyyy-MM-dd HH:mm:ss')),
    tags: [],
    visitsCount: 45,
    visitsSummary: {
      total: 45,
      nonBots: 40,
      bots: 5,
    },
    domain: null,
    meta: {
      validSince: null,
      validUntil: null,
      maxVisits: null,
    },
  };
  const ShortUrlsRow = ShortUrlsRowFactory(fromPartial({
    ShortUrlsRowMenu: () => <span>ShortUrlsRowMenu</span>,
    ColorGenerator: colorGeneratorMock,
    useTimeoutToggle,
  }));

  const setUp = (
    { title, tags = [], meta = {}, settings = {}, search, hasRedirectRules }: SetUpOptions = {},
  ) => renderWithEvents(
    <MemoryRouter initialEntries={search ? [{ search }] : undefined}>
      <SettingsProvider value={fromPartial(settings)}>
        <Table header={<></>}>
          <ShortUrlsRow
            shortUrl={{ ...shortUrl, title, tags, hasRedirectRules, meta: { ...shortUrl.meta, ...meta } }}
            onTagClick={() => null}
          />
        </Table>
      </SettingsProvider>
    </MemoryRouter>,
  );

  it.each([
    { hasRedirectRules: true },
    { hasRedirectRules: false },
  ])('passes a11y checks', (options) => checkAccessibility(setUp(options)));

  it.each([
    [null, 7],
    [undefined, 7],
    ['The title', 8],
  ])('renders expected amount of columns', (title, expectedAmount) => {
    setUp({ title });
    expect(screen.getAllByRole('cell')).toHaveLength(expectedAmount);
  });

  it('renders date in first column', () => {
    setUp();
    expect(screen.getAllByRole('cell')[0]).toHaveTextContent('2018-05-23 18:30');
  });

  it.each([
    [1, shortUrl.shortUrl],
    [2, shortUrl.longUrl],
  ])('renders expected links on corresponding columns', (colIndex, expectedLink) => {
    setUp();

    const col = screen.getAllByRole('cell')[colIndex];
    const link = col.querySelector('a');

    expect(link).toHaveAttribute('href', expectedLink);
  });

  it.each([
    ['My super cool title', 'My super cool title'],
    [undefined, shortUrl.longUrl],
  ])('renders title when short URL has it', (title, expectedContent) => {
    setUp({ title });

    const titleSharedCol = screen.getAllByRole('cell')[2];

    expect(titleSharedCol.querySelector('a')).toHaveAttribute('href', shortUrl.longUrl);
    expect(titleSharedCol).toHaveTextContent(expectedContent);
  });

  it.each([
    [[], ['No tags']],
    [['nodejs', 'reactjs'], ['nodejs', 'reactjs']],
  ])('renders list of tags in fourth row', (tags, expectedContents) => {
    setUp({ tags });
    const cell = screen.getAllByRole('cell')[3];

    expectedContents.forEach((content) => expect(cell).toHaveTextContent(content));
  });

  it.each([
    [{}, '', shortUrl.visitsSummary?.total],
    [fromPartial<Settings>({ visits: { excludeBots: false } }), '', shortUrl.visitsSummary?.total],
    [fromPartial<Settings>({ visits: { excludeBots: true } }), '', shortUrl.visitsSummary?.nonBots],
    [fromPartial<Settings>({ visits: { excludeBots: false } }), 'excludeBots=true', shortUrl.visitsSummary?.nonBots],
    [fromPartial<Settings>({ visits: { excludeBots: true } }), 'excludeBots=true', shortUrl.visitsSummary?.nonBots],
    [{}, 'excludeBots=true', shortUrl.visitsSummary?.nonBots],
    [fromPartial<Settings>({ visits: { excludeBots: true } }), 'excludeBots=false', shortUrl.visitsSummary?.total],
    [fromPartial<Settings>({ visits: { excludeBots: false } }), 'excludeBots=false', shortUrl.visitsSummary?.total],
    [{}, 'excludeBots=false', shortUrl.visitsSummary?.total],
  ])('renders visits count in fifth row', (settings, search, expectedAmount) => {
    setUp({ settings, search });
    expect(screen.getAllByRole('cell', { hidden: true })[4]).toHaveTextContent(`${expectedAmount}`);
  });

  it.each([
    [{ validUntil: formatISO(subDays(now(), 1)) }, ['fa-calendar-xmark', 'tw:text-danger']],
    [{ validSince: formatISO(addDays(now(), 1)) }, ['fa-calendar-xmark', 'tw:text-warning']],
    [{ maxVisits: 45 }, ['fa-link-slash', 'tw:text-danger']],
    [{ maxVisits: 45, validSince: formatISO(addDays(now(), 1)) }, ['fa-link-slash', 'tw:text-danger']],
    [
      { validSince: formatISO(addDays(now(), 1)), validUntil: formatISO(subDays(now(), 1)) },
      ['fa-calendar-xmark', 'tw:text-danger'],
    ],
    [
      { validSince: formatISO(subDays(now(), 1)), validUntil: formatISO(addDays(now(), 1)) },
      ['fa-check', 'tw:text-brand'],
    ],
    [{ maxVisits: 500 }, ['fa-check', 'tw:text-brand']],
    [{}, ['fa-check', 'tw:text-brand']],
  ])('displays expected status icon', (meta, expectedIconClasses) => {
    setUp({ meta });
    const icons = screen.getAllByRole('img', { hidden: true });
    const statusIcon = icons[icons.length - 1];

    expect(statusIcon).toBeInTheDocument();
    expectedIconClasses.forEach((expectedClass) => expect(statusIcon).toHaveClass(expectedClass));
    expect(statusIcon).toMatchSnapshot();
  });

  it.each([
    { hasRedirectRules: true },
    { hasRedirectRules: false },
  ])('shows indicator when a short URL has redirect rules', ({ hasRedirectRules }) => {
    setUp({ hasRedirectRules });

    if (hasRedirectRules) {
      expect(screen.getByTitle('This short URL has dynamic redirect rules')).toBeInTheDocument();
    } else {
      expect(screen.queryByTitle('This short URL has dynamic redirect rules')).not.toBeInTheDocument();
    }
  });
});
