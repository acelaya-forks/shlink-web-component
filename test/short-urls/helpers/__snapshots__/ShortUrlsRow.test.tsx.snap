// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ShortUrlsRow /> > displays expected status icon 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-calendar-xmark tw:text-danger"
  data-icon="calendar-xmark"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M160 0c17.7 0 32 14.3 32 32V64H320V32c0-17.7 14.3-32 32-32s32 14.3 32 32V64h48c26.5 0 48 21.5 48 48v48H32V112c0-26.5 21.5-48 48-48h48V32c0-17.7 14.3-32 32-32zM32 192H480V464c0 26.5-21.5 48-48 48H80c-26.5 0-48-21.5-48-48V192zM337 305c9.4-9.4 9.4-24.6 0-33.9s-24.6-9.4-33.9 0l-47 47-47-47c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9l47 47-47 47c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l47-47 47 47c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-47-47 47-47z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 2`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-calendar-xmark tw:text-warning"
  data-icon="calendar-xmark"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M160 0c17.7 0 32 14.3 32 32V64H320V32c0-17.7 14.3-32 32-32s32 14.3 32 32V64h48c26.5 0 48 21.5 48 48v48H32V112c0-26.5 21.5-48 48-48h48V32c0-17.7 14.3-32 32-32zM32 192H480V464c0 26.5-21.5 48-48 48H80c-26.5 0-48-21.5-48-48V192zM337 305c9.4-9.4 9.4-24.6 0-33.9s-24.6-9.4-33.9 0l-47 47-47-47c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9l47 47-47 47c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l47-47 47 47c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-47-47 47-47z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 3`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-link-slash tw:text-danger"
  data-icon="link-slash"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 640 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L489.3 358.2l90.5-90.5c56.5-56.5 56.5-148 0-204.5c-50-50-128.8-56.5-186.3-15.4l-1.6 1.1c-14.4 10.3-17.7 30.3-7.4 44.6s30.3 17.7 44.6 7.4l1.6-1.1c32.1-22.9 76-19.3 103.8 8.6c31.5 31.5 31.5 82.5 0 114l-96 96-31.9-25C430.9 239.6 420.1 175.1 377 132c-52.2-52.3-134.5-56.2-191.3-11.7L38.8 5.1zM239 162c30.1-14.9 67.7-9.9 92.8 15.3c20 20 27.5 48.3 21.7 74.5L239 162zM406.6 416.4L220.9 270c-2.1 39.8 12.2 80.1 42.2 110c38.9 38.9 94.4 51 143.6 36.3zm-290-228.5L60.2 244.3c-56.5 56.5-56.5 148 0 204.5c50 50 128.8 56.5 186.3 15.4l1.6-1.1c14.4-10.3 17.7-30.3 7.4-44.6s-30.3-17.7-44.6-7.4l-1.6 1.1c-32.1 22.9-76 19.3-103.8-8.6C74 372 74 321 105.5 289.5l61.8-61.8-50.6-39.9z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 4`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-link-slash tw:text-danger"
  data-icon="link-slash"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 640 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L489.3 358.2l90.5-90.5c56.5-56.5 56.5-148 0-204.5c-50-50-128.8-56.5-186.3-15.4l-1.6 1.1c-14.4 10.3-17.7 30.3-7.4 44.6s30.3 17.7 44.6 7.4l1.6-1.1c32.1-22.9 76-19.3 103.8 8.6c31.5 31.5 31.5 82.5 0 114l-96 96-31.9-25C430.9 239.6 420.1 175.1 377 132c-52.2-52.3-134.5-56.2-191.3-11.7L38.8 5.1zM239 162c30.1-14.9 67.7-9.9 92.8 15.3c20 20 27.5 48.3 21.7 74.5L239 162zM406.6 416.4L220.9 270c-2.1 39.8 12.2 80.1 42.2 110c38.9 38.9 94.4 51 143.6 36.3zm-290-228.5L60.2 244.3c-56.5 56.5-56.5 148 0 204.5c50 50 128.8 56.5 186.3 15.4l1.6-1.1c14.4-10.3 17.7-30.3 7.4-44.6s-30.3-17.7-44.6-7.4l-1.6 1.1c-32.1 22.9-76 19.3-103.8-8.6C74 372 74 321 105.5 289.5l61.8-61.8-50.6-39.9z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 5`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-calendar-xmark tw:text-danger"
  data-icon="calendar-xmark"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M160 0c17.7 0 32 14.3 32 32V64H320V32c0-17.7 14.3-32 32-32s32 14.3 32 32V64h48c26.5 0 48 21.5 48 48v48H32V112c0-26.5 21.5-48 48-48h48V32c0-17.7 14.3-32 32-32zM32 192H480V464c0 26.5-21.5 48-48 48H80c-26.5 0-48-21.5-48-48V192zM337 305c9.4-9.4 9.4-24.6 0-33.9s-24.6-9.4-33.9 0l-47 47-47-47c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9l47 47-47 47c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l47-47 47 47c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-47-47 47-47z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 6`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-check tw:text-brand"
  data-icon="check"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 448 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 7`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-check tw:text-brand"
  data-icon="check"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 448 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"
    fill="currentColor"
  />
</svg>
`;

exports[`<ShortUrlsRow /> > displays expected status icon 8`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-check tw:text-brand"
  data-icon="check"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 448 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"
    fill="currentColor"
  />
</svg>
`;
